File: backend/src/AnimeShop/API/Security/TransportEntries.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Security.TransportEntries where

import AnimeShop.Prelude

type TransportEntriesAPI =
  "employee_entries" :>
  (
    Get '[JSON] [Entity TransportEntry] :<|>
    ReqBody '[JSON] TransportEntry :> Post '[JSON] (Entity TransportEntry) :<|>
    ReqBody '[JSON] TransportEntryId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity TransportEntry) :> Patch '[JSON] ()
  )

transportEntriesServer :: ServerT TransportEntriesAPI App
transportEntriesServer = getTransportEntries :<|> postTransportEntry :<|> deleteTransportEntry :<|> patchTransportEntry

getTransportEntries :: App [Entity TransportEntry]
getTransportEntries = runDb $ selectList [] []

postTransportEntry :: TransportEntry -> App (Entity TransportEntry)
postTransportEntry = runDb . insertEntity

deleteTransportEntry :: TransportEntryId -> App ()
deleteTransportEntry = runDb . delete

patchTransportEntry :: Entity TransportEntry -> App ()
patchTransportEntry entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/security/transportEntries.js
import { ADD_TRANSPORT_ENTRY, GET_TRANSPORT_ENTRIES, DELETE_TRANSPORT_ENTRY } from '../../actions/types';

export function transportEntries(state = [], action) {
  switch (action.type) {
    case ADD_TRANSPORT_ENTRY:
      return [ ...state, action.transportEntry];
    case GET_TRANSPORT_ENTRIES:
      return action.transportEntries;
    case DELETE_TRANSPORT_ENTRY:
      return state.filter(transportEntry => transportEntry.id !== action.id);
    default:
      return state;
  }
}
